File: py_system/ingestion/pdf_processor.py
"""PDF Processing Module for Multimodal RAG"""
import logging
from typing import List, Optional, Union, Dict, Any
from pathlib import Path
from dataclasses import dataclass

logger = logging.getLogger(__name__)

# Try to import PDF processing libraries
try:
    import fitz  # PyMuPDF
    HAS_PYMUPDF = True
except ImportError:
    HAS_PYMUPDF = False

try:
    import pdfplumber
    HAS_PDFPLUMBER = True
except ImportError:
    HAS_PDFPLUMBER = False


class PDFProcessor:
    """Extract text from PDF documents with multiple fallback options."""

    def __init__(self, extract_images=False):
        self.extract_images = extract_images

    def process_pdf(self, pdf_path: Union[str, Path]) -> str:
        """Extract text from PDF file."""
        pdf_path = Path(pdf_path)
        if not pdf_path.exists():
            raise FileNotFoundError(f"PDF file not found: {pdf_path}")

        # Try PyMuPDF first (fastest and best quality)
        if HAS_PYMUPDF:
            return self._extract_pymupdf(pdf_path)

        # Fallback to pdfplumber
        if HAS_PDFPLUMBER:
            return self._extract_pdfplumber(pdf_path)

        # Last resort: use command line tool
        return self._extract_pdftotext(pdf_path)

    def _extract_pymupdf(self, pdf_path: Path) -> str:
        """Extract using PyMuPDF (fitz)."""
        text_parts = []
        doc = fitz.open(pdf_path)
        for page_num in range(len(doc)):
            page = doc[page_num]
            text = page.get_text("text")
            if text.strip():
                text_parts.append(f"--- Page {page_num + 1} ---\n{text}")
        doc.close()
        return "\n\n".join(text_parts)

    def _extract_pdfplumber(self, pdf_path: Path) -> str:
        """Extract using pdfplumber (good for tables)."""
        text_parts = []
        with pdfplumber.open(pdf_path) as pdf:
            for page_num, page in enumerate(pdf.pages):
                text = page.extract_text()
                if text:
                    text_parts.append(f"--- Page {page_num + 1} ---\n{text}")

                # Extract tables if present
                tables = page.extract_tables()
                for table_idx, table in enumerate(tables):
                    if table:
                        table_text = self._format_table(table)
                        text_parts.append(f"--- Table {table_idx + 1} ---\n{table_text}")

        return "\n\n".join(text_parts)

    def _extract_pdftotext(self, pdf_path: Path) -> str:
        """Fallback: use pdftotext command line tool."""
        import subprocess
        temp_file = str(pdf_path) + ".txt"
        try:
            result = subprocess.run(
                ["pdftotext", "-layout", str(pdf_path), temp_file],
                capture_output=True,
                text=True
            )
            if Path(temp_file).exists():
                with open(temp_file, 'r', encoding='utf-8') as f:
                    text = f.read()
                Path(temp_file).unlink()
                return text
        except FileNotFoundError:
            pass

        raise RuntimeError("No PDF extraction method available. Install pymupdf or pdfplumber.")

    def _format_table(self, table: list) -> str:
        """Format table data as readable text."""
        lines = []
        for row in table:
            if row:
                cleaned = [cell.strip() if cell else "" for cell in row]
                lines.append(" | ".join(cleaned))
        return "\n".join(lines)

    def extract_pages(self, pdf_path: Union[str, Path]) -> List[Dict[str, Any]]:
        """Extract text page by page."""
        pdf_path = Path(pdf_path)
        if not pdf_path.exists():
            raise FileNotFoundError(f"PDF file not found: {pdf_path}")

        pages = []
        if HAS_PYMUPDF:
            doc = fitz.open(pdf_path)
            for page_num in range(len(doc)):
                page = doc[page_num]
                text = page.get_text("text")
                pages.append({
                    'page_num': page_num + 1,
                    'text': text.strip()
                })
            doc.close()
        elif HAS_PDFPLUMBER:
            with pdfplumber.open(pdf_path) as pdf:
                for page_num, page in enumerate(pdf.pages):
                    text = page.extract_text()
                    pages.append({
                        'page_num': page_num + 1,
                        'text': text.strip() if text else ""
                    })
        else:
            # Fallback to single extraction
            pages.append({'page_num': 1, 'text': self.process_pdf(pdf_path)})

        return pages


class PDFIngester:
    def __init__(self, pdf_processor: Optional[PDFProcessor] = None):
        self.pdf_processor = pdf_processor or PDFProcessor()

    def ingest_pdf(self, pdf_path: Union[str, Path], source_name: Optional[str] = None) -> dict:
        """Ingest a PDF file and return extracted text with metadata."""
        pdf_path = Path(pdf_path)
        source_name = source_name or pdf_path.name

        extracted_text = self.pdf_processor.process_pdf(pdf_path)
        pages = self.pdf_processor.extract_pages(pdf_path)

        return {
            "text": extracted_text,
            "page_count": len(pages),
            "metadata": {
                "source": source_name,
                "type": "pdf",
                "filename": pdf_path.name,
                "pages": pages
            }
        }

    def ingest_pdfs_from_directory(self, directory: Union[str, Path], extensions=None) -> List[dict]:
        """Ingest all PDF files from a directory."""
        directory = Path(directory)
        extensions = extensions or [".pdf"]

        results = []
        for ext in extensions:
            for pdf_path in directory.glob(f"*{ext}"):
                try:
                    result = self.ingest_pdf(pdf_path)
                    results.append(result)
                except Exception as e:
                    logger.error(f"Failed to ingest {pdf_path}: {e}")

        logger.info(f"Ingested {len(results)} PDFs from {directory}")
        return results 

File: src/ingestion/pdf-processor.js
/**
 * PDF Processing Module for Multimodal RAG
 * Extracts text from PDF documents with multiple fallback options
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { Logger } from '../utils/index.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const logger = new Logger({ name: 'PDFProcessor', level: process.env.RAG_LOG_LEVEL || process.env.LOG_LEVEL || 'info' });

// Try to import PDF processing libraries
let pdfParse = null;
let pdfLib = null;
let pdfjsLib = null;

let HAS_PDFPARSE = false;
let HAS_PDFLIB = false;
let HAS_PDFJS = false;

try {
    const pdfParseModule = await import('pdf-parse');
    pdfParse = pdfParseModule?.PDFParse || pdfParseModule?.default?.PDFParse || null;
    HAS_PDFPARSE = typeof pdfParse === 'function';
    if (!HAS_PDFPARSE) {
        logger.warn('pdf-parse loaded but no PDFParse class export found');
    }
} catch (e) {
    logger.warn('pdf-parse not available', { error: e.message });
}

try {
    pdfjsLib = await import('pdfjs-dist/legacy/build/pdf.mjs');
    HAS_PDFJS = true;
} catch (e) {
    logger.warn('pdfjs-dist not available', { error: e.message });
}

class PDFProcessor {
    /**
     * @param {Object} options - Options for PDF processing
     * @param {boolean} options.extractImages - Whether to extract images
     */
    constructor(options = {}) {
        this.extractImages = options.extractImages || false;
    }

    /**
     * Extract text from PDF file
     * @param {string} pdfPath - Path to PDF file
     * @returns {Promise<string>} Extracted text
     */
    async processPdf(pdfPath) {
        if (!fs.existsSync(pdfPath)) {
            throw new Error(`PDF file not found: ${pdfPath}`);
        }

        // Try pdf-parse first (simple and reliable)
        if (HAS_PDFPARSE) {
            return await this._extractPdfParse(pdfPath);
        }

        // Fallback to pdf.js
        if (HAS_PDFJS) {
            return await this._extractPdfJs(pdfPath);
        }

        // Last resort: use pdf-lib
        if (HAS_PDFLIB) {
            return await this._extractPdfLib(pdfPath);
        }

        throw new Error('No PDF extraction method available. Install pdf-parse, pdfjs-dist, or pdf-lib.');
    }

    /**
     * Extract using pdf-parse
     * @param {string} pdfPath - Path to PDF file
     * @returns {Promise<string>} Extracted text
     */
    async _extractPdfParse(pdfPath) {
        const dataBuffer = fs.readFileSync(pdfPath);
        const parser = new pdfParse({ data: dataBuffer });
        try {
            const textResult = await parser.getText();
            return textResult?.text || '';
        } finally {
            try {
                await parser.destroy();
            } catch (e) {
                // ignore cleanup errors
            }
        }
    }

    /**
     * Extract using pdf.js
     * @param {string} pdfPath - Path to PDF file
     * @returns {Promise<string>} Extracted text
     */
    async _extractPdfJs(pdfPath) {
        const dataBuffer = fs.readFileSync(pdfPath);
        const loadingTask = pdfjsLib.getDocument({ data: dataBuffer });
        const pdfDocument = await loadingTask.promise;

        const textParts = [];
        const numPages = pdfDocument.numPages;

        for (let pageNum = 1; pageNum <= numPages; pageNum++) {
            const page = await pdfDocument.getPage(pageNum);
            const textContent = await page.getTextContent();
            const pageText = textContent.items.map(item => item.str).join(' ');
            if (pageText.trim()) {
                textParts.push(`--- Page ${pageNum} ---\n${pageText}`);
            }
        }

        return textParts.join('\n\n');
    }

    /**
     * Extract using pdf-lib
     * @param {string} pdfPath - Path to PDF file
     * @returns {Promise<string>} Extracted text
     */
    async _extractPdfLib(pdfPath) {
        if (!HAS_PDFLIB) {
            try {
                pdfLib = await import('pdf-lib');
                HAS_PDFLIB = true;
            } catch (e) {
                logger.warn('pdf-lib not available', { error: e.message });
                throw e;
            }
        }

        const pdfBytes = fs.readFileSync(pdfPath);
        const pdfDocument = await pdfLib.PDFDocument.load(pdfBytes);
        const pages = pdfDocument.getPages();

        const textParts = [];
        for (let i = 0; i < pages.length; i++) {
            const page = pages[i];
            const { height } = page.getSize();
            const text = page.getTextContent();
            const pageText = text.items.map(item => item.str).join(' ');
            if (pageText.trim()) {
                textParts.push(`--- Page ${i + 1} ---\n${pageText}`);
            }
        }

        return textParts.join('\n\n');
    }

    /**
     * Extract text page by page
     * @param {string} pdfPath - Path to PDF file
     * @returns {Promise<Array>} Array of page objects
     */
    async extractPages(pdfPath) {
        if (!fs.existsSync(pdfPath)) {
            throw new Error(`PDF file not found: ${pdfPath}`);
        }

        const pages = [];

        if (HAS_PDFJS) {
            const dataBuffer = fs.readFileSync(pdfPath);
            const loadingTask = pdfjsLib.getDocument({ data: dataBuffer });
            const pdfDocument = await loadingTask.promise;
            const numPages = pdfDocument.numPages;

            for (let pageNum = 1; pageNum <= numPages; pageNum++) {
                const page = await pdfDocument.getPage(pageNum);
                const textContent = await page.getTextContent();
                const pageText = textContent.items.map(item => item.str).join(' ');
                pages.push({
                    page_num: pageNum,
                    text: pageText.trim()
                });
            }
        } else if (HAS_PDFPARSE) {
            const dataBuffer = fs.readFileSync(pdfPath);
            const parser = new pdfParse({ data: dataBuffer });
            try {
                const textResult = await parser.getText();
                const parsedPages = Array.isArray(textResult?.pages) ? textResult.pages : [];
                if (parsedPages.length > 0) {
                    for (const page of parsedPages) {
                        pages.push({
                            page_num: page.num || pages.length + 1,
                            text: String(page.text || '').trim()
                        });
                    }
                } else {
                    const text = textResult?.text || '';
                    pages.push({ page_num: 1, text });
                }
            } finally {
                try {
                    await parser.destroy();
                } catch (e) {
                    // ignore cleanup errors
                }
            }
        } else {
            // Fallback to single extraction
            const text = await this.processPdf(pdfPath);
            pages.push({ page_num: 1, text });
        }

        return pages;
    }

    /**
     * Format table data as readable text
     * @param {Array} table - Table data
     * @returns {string} Formatted table
     */
    _formatTable(table) {
        const lines = [];
        for (const row of table) {
            if (row) {
                const cleaned = row.map(cell => (cell ? cell.trim() : ''));
                lines.push(cleaned.join(' | '));
            }
        }
        return lines.join('\n');
    }
}

class PDFIngester {
    /**
     * @param {PDFProcessor} pdfProcessor - Optional PDF processor instance
     */
    constructor(pdfProcessor = null) {
        this.pdfProcessor = pdfProcessor || new PDFProcessor();
    }

    /**
     * Ingest a PDF file and return extracted text with metadata
     * @param {string} pdfPath - Path to PDF file
     * @param {string} sourceName - Optional source name
     * @returns {Promise<Object>} Ingested document
     */
    async ingestPdf(pdfPath, sourceName = null) {
        const fileName = path.basename(pdfPath);
        sourceName = sourceName || fileName;

        const extractedText = await this.pdfProcessor.processPdf(pdfPath);
        const pages = await this.pdfProcessor.extractPages(pdfPath);

        return {
            text: extractedText,
            page_count: pages.length,
            metadata: {
                source: sourceName,
                type: 'pdf',
                filename: fileName,
                pages: pages
            }
        };
    }

    /**
     * Ingest all PDF files from a directory
     * @param {string} directory - Directory path
     * @param {Array} extensions - File extensions to look for
     * @returns {Promise<Array>} Array of ingested documents
     */
    async ingestPdfsFromDirectory(directory, extensions = null) {
        extensions = extensions || ['.pdf'];

        const results = [];
        const files = fs.readdirSync(directory);

        for (const file of files) {
            const ext = path.extname(file).toLowerCase();
            if (extensions.includes(ext)) {
                const pdfPath = path.join(directory, file);
                try {
                    const result = await this.ingestPdf(pdfPath);
                    results.push(result);
                } catch (e) {
                    logger.error('Failed to ingest PDF', { path: pdfPath, error: e.message });
                }
            }
        }

        logger.info('Ingested PDFs from directory', { directory, count: results.length });
        return results;
    }
}

export { PDFProcessor, PDFIngester };
export default { PDFProcessor, PDFIngester };
